File: Trabalho 1/simulador_jogo.py
#!/usr/bin/env python3
"""
Simulador de Jogo Coreano - Trabalho de Métodos Numéricos

Este programa calcula a população estacionária de atores em um jogo
onde eles se movem entre n salas conectadas, com probabilidades específicas
de movimento e possibilidade de morte nas extremidades.

Autor: Implementação baseada no enunciado do Trabalho II
"""

import numpy as np
import sys


def criar_matriz_transicao(n):
    """
    Cria a matriz de transição P para n salas.
    
    Args:
        n (int): Número de salas (deve ser ímpar)
    
    Returns:
        numpy.ndarray: Matriz de transição P de tamanho n×n
    """
    print(f"\n=== CRIANDO MATRIZ DE TRANSIÇÃO PARA {n} SALAS ===")
    
    # Probabilidade de ficar na mesma sala
    prob_ficar = 1.0 / n
    print(f"Probabilidade de ficar na mesma sala: {prob_ficar:.4f}")
    
    # Probabilidade total de se mover
    prob_mover = 1.0 - prob_ficar
    print(f"Probabilidade total de movimento: {prob_mover:.4f}")
    
    # Inicializar matriz com zeros
    P = np.zeros((n, n))
    
    # Preencher a matriz sala por sala
    for i in range(n):
        sala_atual = i + 1  # Salas numeradas de 1 a n
        print(f"\nSala {sala_atual}:")
        
        # Probabilidade de ficar na mesma sala
        P[i, i] = prob_ficar
        print(f"  Ficar na sala {sala_atual}: {prob_ficar:.4f}")
        
        # Identificar salas vizinhas
        vizinhas = []
        if i > 0:  # Tem vizinha à esquerda
            vizinhas.append(i - 1)
        if i < n - 1:  # Tem vizinha à direita
            vizinhas.append(i + 1)
        
        print(f"  Salas vizinhas: {[v+1 for v in vizinhas]}")
        
        # Distribuir probabilidade de movimento entre vizinhas
        if len(vizinhas) > 0:
            # Para salas das pontas: metade da prob_mover vai para vizinha, metade é morte
            # Para salas internas: prob_mover é dividida igualmente entre as 2 vizinhas
            if len(vizinhas) == 1:  # Sala da ponta
                prob_por_vizinha = prob_mover / 2.0  # Metade para vizinha
                prob_morte = prob_mover / 2.0        # Metade para morte
                print(f"  Probabilidade para vizinha: {prob_por_vizinha:.4f}")
                print(f"  Probabilidade de morte: {prob_morte:.4f}")
            else:  # Sala interna
                prob_por_vizinha = prob_mover / 2.0  # Dividir igualmente entre 2 vizinhas
                prob_morte = 0.0
                print(f"  Probabilidade por vizinha: {prob_por_vizinha:.4f}")
            
            for j in vizinhas:
                P[i, j] = prob_por_vizinha
                print(f"  Ir para sala {j+1}: {prob_por_vizinha:.4f}")
        
        # Mostrar probabilidade de morte para referência
        if len(vizinhas) == 1:
            print(f"  💀 Morte (sair do jogo): {prob_morte:.4f}")
    
    print(f"\nMatriz de transição P ({n}×{n}):")
    print(P)
    
    # Verificar se as linhas somam corretamente
    print(f"\nVerificação - soma das linhas (deve ser ≤ 1.0):")
    for i in range(n):
        soma = np.sum(P[i, :])
        prob_morte = 1.0 - soma
        print(f"  Sala {i+1}: soma = {soma:.4f}, morte = {prob_morte:.4f}")
    
    return P


def calcular_populacao_estacionaria(n, taxa_chegada=1.0):
    """
    Calcula a distribuição estacionária de população nas salas.
    
    Args:
        n (int): Número de salas
        taxa_chegada (float): Taxa de chegada de novos atores por episódio
    
    Returns:
        numpy.ndarray: Vetor com população esperada em cada sala
    """
    print(f"\n=== CALCULANDO POPULAÇÃO ESTACIONÁRIA ===")
    print(f"Número de salas: {n}")
    print(f"Taxa de chegada: {taxa_chegada} atores/episódio")
    
    # Criar matriz de transição
    P = criar_matriz_transicao(n)
    
    # Criar matriz identidade
    I = np.eye(n)
    print(f"\nMatriz identidade I ({n}×{n}) criada")
    
    # Calcular (I - P^T)
    A = I - P.T
    print(f"\nMatriz A = I - P^T:")
    print(A)
    
    # Criar vetor de chegadas (b)
    b = np.zeros(n)
    sala_central = (n + 1) // 2 - 1  # Índice da sala central (base 0)
    b[sala_central] = taxa_chegada
    
    print(f"\nSala central: {sala_central + 1} (índice {sala_central})")
    print(f"Vetor de chegadas b: {b}")
    
    # Resolver sistema linear A × x = b
    print(f"\nResolvendo sistema linear A × x = b...")
    try:
        x = np.linalg.solve(A, b)
        print("Sistema resolvido com sucesso!")
    except np.linalg.LinAlgError as e:
        print(f"Erro ao resolver sistema: {e}")
        return None
    
    print(f"\nSolução x (população por sala):")
    for i in range(n):
        print(f"  Sala {i+1}: {x[i]:.6f} atores")
    
    return x


def main():
    """
    Função principal do programa.
    """
    print("=" * 60)
    print("SIMULADOR DE JOGO COREANO")
    print("Trabalho de Métodos Numéricos")
    print("=" * 60)
    
    # Verificar argumentos da linha de comando
    if len(sys.argv) != 2:
        print("\nUso: python simulador_jogo.py <número_de_salas>")
        print("Exemplo: python simulador_jogo.py 23")
        
        # Executar exemplos padrão
        print("\nExecutando exemplos padrão...")
        numeros_salas = [3, 5, 7, 23]
    else:
        try:
            n = int(sys.argv[1])
            if n <= 0 or n % 2 == 0:
                print("Erro: O número de salas deve ser um inteiro positivo ímpar!")
                sys.exit(1)
            numeros_salas = [n]
        except ValueError:
            print("Erro: Por favor, forneça um número inteiro válido!")
            sys.exit(1)
    
    # Executar simulação para cada número de salas
    for n in numeros_salas:
        print(f"\n{'='*60}")
        print(f"SIMULAÇÃO PARA {n} SALAS")
        print(f"{'='*60}")
        
        # Calcular população estacionária
        x = calcular_populacao_estacionaria(n, taxa_chegada=1.0)
        
        if x is not None:
            populacao_total = np.sum(x)
            print(f"\n🎯 RESULTADO FINAL:")
            print(f"População para {n} salas: {populacao_total:.0f} atores")
            print(f"População total exata: {populacao_total:.6f}")
            
            # Mostrar distribuição detalhada
            print(f"\nDistribuição por sala:")
            for i in range(n):
                percentual = (x[i] / populacao_total) * 100
                print(f"  Sala {i+1}: {x[i]:8.4f} atores ({percentual:5.1f}%)")
            
            # Análise adicional
            print(f"\nAnálise:")
            sala_central = (n + 1) // 2
            print(f"  - Sala central ({sala_central}): {x[sala_central-1]:.4f} atores")
            print(f"  - Sala com mais atores: Sala {np.argmax(x)+1} ({np.max(x):.4f})")
            print(f"  - Sala com menos atores: Sala {np.argmin(x)+1} ({np.min(x):.4f})")


if __name__ == "__main__":
    main()

File: Trabalho 1/simulador_jogo.cpp
// Trabalho 1 - Simulador de Jogo Coreano
// Bernardo Klein Heitz 
// Lucas Langer Lantmann

//# para compilar:
//      g++ -o simulador simulador_jogo.cpp

//# executar com diferentes numeros de salas (sempre numeros inteiros impares!)
//      ./simulador 3
//      ./simulador 5
//      ./simulador 7
//      ./simulador 23
//  ou  simulador.exe 3


#include <iostream> // biblioteca para entrada e saída de dados
#include <vector> // uso de vetores
#include <iomanip> // formatação de saída
#include <cmath> 
#include <algorithm> // usar algoritmos

using namespace std; 


class SimuladorJogo { // classe para simular o jogo

private: 
    int n;  // numero de salas
    vector<vector<double>> P;  // matriz de transição
    vector<double> x;  // populacao por sala


    void criarMatrizTransicao() { // função para criar a matriz de transição
        cout << "\n======CRIANDO MATRIZ DE TRANSICAO PARA " << n << " SALAS======\n";
        
        double prob_ficar = 1.0 / n; // prob de ficar na mesma sala
        double prob_mover = 1.0 - prob_ficar; // prob de mover para uma sala vizinha
        
        cout << fixed << setprecision(4); // formata a saida para 4 casas decimais
        cout << "prob de ficar na mesma sala: " << prob_ficar << "\n";
        cout << "prob total de movimento: " << prob_mover << "\n";
        
        // Inicializar matriz com zeros
        P.assign(n, vector<double>(n, 0.0));
        
        for (int i = 0; i < n; i++) { // percorre as salas
            int sala_atual = i + 1; // indice da sala atual
            cout << "\nsala " << sala_atual << ":\n"; // mostra a sala atual
            
            // prob de ficar na mesma sala
            P[i][i] = prob_ficar;
            cout << " ficar na sala " << sala_atual << ": " << prob_ficar << "\n"; // mostra a prob de ficar na sala atual
            
            // Identificar salas vizinhas
            vector<int> vizinhas;
            if (i > 0) vizinhas.push_back(i - 1); // adiciona a sala anterior
            if (i < n - 1) vizinhas.push_back(i + 1); // adiciona a sala posterior
            
            cout << "  salas vizinhas: [";
            for (size_t j = 0; j < vizinhas.size(); j++) {
                cout << (vizinhas[j] + 1); // mostra a sala vizinha
                if (j < vizinhas.size() - 1) cout << ", "; // mostra a sala vizinha
            }
            cout << "]\n"; 
            
            // Distribuir prob de movimento
            if (!vizinhas.empty()) { // se a sala tiver vizinhas
                double prob_por_vizinha; 
                double prob_morte = 0.0;
                
                if (vizinhas.size() == 1) {  // sala da ponta
                    prob_por_vizinha = prob_mover / 2.0; // prob de mover para a vizinha
                    prob_morte = prob_mover / 2.0; // prob de morrer
                    cout << "  prob para vizinha: " << prob_por_vizinha << "\n"; // mostra a prob de mover para a vizinha
                    cout << "  prob de morte: " << prob_morte << "\n"; // mostra a prob de morrer
                } else {  // Sala interna
                    prob_por_vizinha = prob_mover / 2.0; // prob de mover para a vizinha
                    cout << "  prob por vizinha: " << prob_por_vizinha << "\n"; // mostra a prob de mover para a vizinha
                }
                
                for (int j : vizinhas) { // percorre as salas vizinhas
                    P[i][j] = prob_por_vizinha;
                    cout << "  ir para sala " << (j + 1) << ": " << prob_por_vizinha << "\n";
                }
                
                if (vizinhas.size() == 1) { // se a sala tiver apenas uma vizinha
                    cout << "  morte! (sair do jogo): " << prob_morte << "\n";
                }
            }
        }
    }
    

    void resolverSistemaLinear() { // funcao para resolver o sistema linear (Ax = b)
        cout << "\n=====RESOLVENDO SISTEMA LINEAR=====\n";
        
        // criar matriz A = I - P^T
        vector<vector<double>> A(n, vector<double>(n, 0.0)); // matriz A (n x n) 
        for (int i = 0; i < n; i++) { // percorre as linhas
            for (int j = 0; j < n; j++) { // percorre as colunas
                A[i][j] = (i == j ? 1.0 : 0.0) - P[j][i];  // P^T (matriz transposta) 
            }
        }
        
            // criar vetor b (chegadas)
        vector<double> b(n, 0.0); // vetor b (n x 1)
        int sala_central = (n + 1) / 2 - 1;  // indice base 0
        b[sala_central] = 1.0;  // taxa de chegada

        cout << "sala central: " << (sala_central + 1) << " (indice " << sala_central << ")\n"; // indice base 0
        
        // resolver Ax = b usando eliminacao gaussiana
        x = resolverGaussiana(A, b); // chama a funcao resolverGaussiana
    }
    
    vector<double> resolverGaussiana(vector<vector<double>> A, vector<double> b) { // funcao para resolver o sistema linear
        int n = A.size(); // tamanho da matriz A
    
        // eliminacao para frente
        for (int i = 0; i < n; i++) {
            // encontrar pivô
            int max_row = i; // indice da linha com o maior valor absoluto
            for (int k = i + 1; k < n; k++) { // percorre as linhas abaixo da linha i
                if (abs(A[k][i]) > abs(A[max_row][i])) { // abs = valor absoluto
                    max_row = k; // atualiza o indice da linha com o maior valor absoluto
                }
            }
            
            // trocar linhas
            swap(A[i], A[max_row]); // troca as linhas i e max_row
            swap(b[i], b[max_row]); // troca os elementos i e max_row
            
            // tornar diagonal principal 1
            double pivot = A[i][i]; // elemento da diagonal principal
            if (abs(pivot) < 1e-10) { // abs = valor absoluto
                throw runtime_error("matriz singular (o sistema nao pode ser resolvido)"); // matriz singular
            }
            
            for (int k = i; k < n; k++) { // percorre as colunas
                A[i][k] /= pivot; // divide a linha i pelo elemento da diagonal principal
            }
            b[i] /= pivot; // divide o elemento i pelo elemento da diagonal principal
            
            // eliminacao
            for (int k = i + 1; k < n; k++) { // percorre as linhas abaixo da linha i
                double factor = A[k][i]; // elemento da linha k e coluna i
                for (int j = i; j < n; j++) { // percorre as colunas
                    A[k][j] -= factor * A[i][j]; // subtrai o produto do elemento da linha k e coluna j pelo elemento da linha i e coluna j
                }
                b[k] -= factor * b[i]; // subtrai o produto do elemento da linha k pelo elemento da linha i
            }
        }
        
        // substituicao para tras
        vector<double> x(n); // vetor x (n x 1)
        for (int i = n - 1; i >= 0; i--) { // percorre as linhas de baixo para cima
            x[i] = b[i]; // atribui o valor de b[i] para x[i]
            for (int j = i + 1; j < n; j++) { 
                x[i] -= A[i][j] * x[j]; // subtrai o produto do elemento da linha i e coluna j pelo elemento da linha i e coluna j
            }
        }
        
        return x; // retorna o vetor x (solucao do sistema linear)
    }


public: 
    SimuladorJogo(int num_salas) : n(num_salas) { // construtor da classe SimuladorJogo (parametro num_salas) 
        if (n <= 0 || n % 2 == 0) { // se o numero de salas for menor ou igual a 0 ou for par
            throw invalid_argument("n de salas deve ser um inteiro positivo impar!"); // lanca uma excecao
        }
    }
    
    void executar() { // funcao para executar o programa
        cout << "==================================\n";
        cout << "SIMULADOR DE JOGO COREANO\n";
        cout << "==================================\n";
        cout << "simulacao para " << n << " salas\n";
        
        try {
            criarMatrizTransicao(); // chama a funcao criarMatrizTransicao
            resolverSistemaLinear(); // chama a funcao resolverSistemaLinear
            mostrarResultados(); // chama a funcao mostrarResultados
        } catch (const exception& e) {
            cout << "erro: " << e.what() << "\n"; 
        }
    }
    

    void mostrarResultados() { // funcao para mostrar os resultados
        cout << "\n===RESULT FINAL=== :\n";
        
        double populacao_total = 0.0; // populacao total
        for (double pop : x) {
            populacao_total += pop; // soma a populacao de cada sala
        }
        
        cout << fixed << setprecision(0); // formata a saida para 0 casas decimais
        cout << "populacao para " << n << " salas: " << round(populacao_total) << " atores\n";
        
        cout << fixed << setprecision(6); // formata a saida para 6 casas decimais
        cout << "populacao total exata: " << populacao_total << "\n";
        
        cout << "\ndistribuicao por sala:\n";
        cout << fixed << setprecision(4); // formata a saida para 4 casas decimais
        for (int i = 0; i < n; i++) { // percorre as salas
            double percentual = (x[i] / populacao_total) * 100.0; // calcula o percentual de atores em cada sala
            cout << "  sala " << setw(2) << (i + 1) << ": " 
                 << setw(8) << x[i] << " atores (" 
                 << setw(5) << setprecision(1) << percentual << "%)\n";
        }
        
        // analise
        cout << "\nanalise:\n";
        int sala_central = (n + 1) / 2; // indice da sala central
        cout << "  sala central (" << sala_central << "): " 
             << fixed << setprecision(4) << x[sala_central - 1] << " atores\n";
        
        auto max_it = max_element(x.begin(), x.end()); // encontra o maior elemento do vetor x
        auto min_it = min_element(x.begin(), x.end()); // encontra o menor elemento do vetor x
        
        cout << "  sala com + atores: sala numero "  // sala com mais atores
             << (max_it - x.begin() + 1) << " (" << *max_it << ")\n";
        cout << "  sala com - atores: sala numero "  // sala com menos atores
             << (min_it - x.begin() + 1) << " (" << *min_it << ")\n";
    }
};


int main(int argc, char* argv[]) { // funcao principal (argc = numero de argumentos, argv = vetor de argumentos)
    if (argc != 2) { // se o numero de argumentos for diferente de 2
        cout << "uso: " << argv[0] << " <numero_de_salas>\n"; // mostra o uso do programa
        cout << "exemplo: " << argv[0] << " 23\n"; // mostra um exemplo de uso do programa
        return 1; // retorna 1 se o programa for executado com erro
    }
    
    try { // try = tenta executar o codigo
        int n = stoi(argv[1]); // converte o primeiro argumento para um inteiro
        SimuladorJogo simulador(n); // cria um objeto da classe SimuladorJogo
        simulador.executar(); // chama a funcao executar
    } catch (const exception& e) {
        cout << "erro: " << e.what() << "\n"; 
        return 1; // retorna 1 se o programa for executado com erro
    }
    
    return 0; // retorna 0 se o programa for executado com sucesso
}
